### DISCUSSION TITLE
SoloV2 on Ubuntu

### DISCUSSION THREAD

**Discussion by jaclas** [solokeys/solo2#132] [Q&A] [answered]
I don't understand why only Windows and macOS appear in the documentation and pages? Please also consider Linux users, especially those who are not fluent in the console. 

I am using Ubuntu 22.04, how do I do the upgrade?

**Comment by needs-coffee** [comment, marked as answer]
@jaclas 
Obvious caveat that you may loose access to credentials on the device so have a backup method to access your accounts before starting (should have one anyway). 
The update is installed with the [solo2-cli application](https://github.com/solokeys/solo2-cli)
2 paths you can choose - use the precompiled binary for the solo2-cli from github releases, or install with cargo.
Below is very similar to the MacOS process and was tested on ubuntu 22.04.

## install with github releases
1. Go to https://github.com/solokeys/solo2-cli
2. go to releases and download the latest binary for your platform - in your case [solo2-v0.2.0-x86_64-unknown-linux-gnu](https://github.com/solokeys/solo2-cli/releases/download/v0.2.0/solo2-v0.2.0-x86_64-unknown-linux-gnu)
3. open a new terminal and cd to where the binary was downloaded - by default this should be ` cd ~/Downloads`
4. Rename it to a more friendly name - `mv solo2-v0.2.0-x86_64-unknown-linux-gnu solo2`
5. Give the downloaded binary execute permissions - ` chmod +x solo2`
6. Insert your solokey
7. run `./solo2 ls` - this should list the connected keys and current firmware
8. run the update with `sudo ./solo2 update` - this will download and update to the latest firmware. To Run this update without requiring sudo install the [udev rules for the LPC55 bootloader](https://github.com/solokeys/solo2-cli/blob/main/70-solo2.rules)
9. Run the ls command again to confirm the update was successful. 

## install with cargo
1. Install rust 
a. Open a terminal window
b. Update your system `sudo apt update && sudo apt upgrade`
c. run `curl https://sh.rustup.rs -sSf | sh`
d. read the prompt and when ready proceed with option 1. This takes a minute - go make a cuppa
e. when done - restart your terminal (otherwise the rust commands wont be found)
f. check rust and cargo installed correctly with `rustc --verison` and `cargo --verison` - both should print version info.
2. Install the required libraries - `sudo apt update && sudo apt install libpcsclite-dev  libudev-dev`
3. install the solo2 client - ` cargo install solo2` 
5. continue from step 6 above
6. If you have not installed the udev rules for the LPC55 bootlaoder you will need to run the update commands as root. As solo2 has been installed with cargo locally to your user you will need to specify the full path to the binary file when run as root. To do this as one command run `sudo $(which solo2) update`

EDIT
edited to reflect the need for sudo or udev rules as per nickray below

**Reply by nickray** [reply]
Regarding sudo, you need to install the https://github.com/solokeys/solo2-cli/blob/main/70-solo2.rules udev rules to be able to talk to the device in maintenance mode. Basically, `solo2 app admin maintenance` (which is one step of `solo2 update`) switches the device from "Solo 2" mode (USB vendor ID 0x1209, product ID 0xbeee) to "LPC 55" mode (vendor 0x1209, product 0xb000), at which point you're (trying to) talking to the chip's signed firmware update mechanism over USB HID (custom NXP protocol).

It used to be the case that you needed the udev rules also for Solo 2 mode, but systemd added these for all FIDO2 authenticators.

I wonder if there's a way to probe udev rules and display a helpful error message.
